File: backend/app/api/v1/recommendations.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session
from typing import List, Optional

from app.db.session import get_db
from app.schemas.recommendation import (
    RecommendationResponse,
    RecommendationStatusUpdate,
    RecommendationGenerationSummary
)
from app.services.recommendation_service import (
    generate_recommendations,
    list_recommendations,
    get_recommendation,
    update_recommendation_status
)

router = APIRouter(prefix="/recommendations", tags=["Smart Equipment Recommendations"])

@router.get("", response_model=List[RecommendationResponse])
def get_recommendations_endpoint(
    equipment_id: Optional[str] = Query(None, description="Filter by equipment ID"),
    status: Optional[str] = Query(None, description="Filter by status (PENDING, ACCEPTED, DISMISSED)"),
    priority: Optional[str] = Query(None, description="Filter by priority (LOW, MEDIUM, HIGH, CRITICAL)"),
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    db: Session = Depends(get_db)
):
    """
    Query smart equipment recommendations for customer rental optimization.
    """
    return list_recommendations(
        db=db,
        equipment_id=equipment_id,
        status=status,
        priority=priority,
        limit=limit,
        offset=offset
    )

@router.post("/generate", response_model=RecommendationGenerationSummary, status_code=status.HTTP_201_CREATED)
def generate_recommendations_endpoint(
    db: Session = Depends(get_db)
):
    """
    Idempotent recommendation engine:
    Cross-references underutilized rented machinery with site demand forecasts to produce explainable actions.
    """
    return generate_recommendations(db)

@router.get("/{recommendation_id}", response_model=RecommendationResponse)
def get_single_recommendation_endpoint(
    recommendation_id: str,
    db: Session = Depends(get_db)
):
    """
    Retrieve an individual recommendation by ID.
    """
    rec = get_recommendation(db, recommendation_id)
    if not rec:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Recommendation with ID '{recommendation_id}' not found."
        )
    return rec

@router.patch("/{recommendation_id}", response_model=RecommendationResponse)
def update_recommendation_endpoint(
    recommendation_id: str,
    payload: RecommendationStatusUpdate,
    db: Session = Depends(get_db)
):
    """
    Record customer manager decision (ACCEPT or DISMISS) without automatically relocating equipment.
    """
    valid_statuses = ["ACCEPTED", "DISMISSED", "PENDING"]
    norm_status = payload.status.upper()
    if norm_status not in valid_statuses:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Status must be one of {valid_statuses}."
        )

    rec = update_recommendation_status(db, recommendation_id, norm_status)
    if not rec:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Recommendation with ID '{recommendation_id}' not found."
        )
    return rec

File: frontend1/src/api/recommendations.ts
import { apiClient } from './client';

export interface RecommendationResponse {
  recommendation_id: string;
  equipment_id: string;
  equipment_type: string | null;
  current_site_id: string | null;
  current_site_name: string | null;
  recommended_site_id: string | null;
  recommended_site_name: string | null;
  recommendation_type: 'REDEPLOY' | 'RETURN_OR_DOWNSIZE' | 'ASSIGN' | 'RETAIN' | 'MONITOR' | string;
  reason: string;
  expected_utilization_gain: number | null;
  priority: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL' | string;
  status: 'PENDING' | 'ACCEPTED' | 'DISMISSED' | string;
  created_at: string;
}

export interface RecommendationGenerationSummary {
  recommendations_created: number;
  recommendations_updated: number;
  total_active_recommendations: number;
  recommendations: RecommendationResponse[];
}

export interface RecommendationFilterParams {
  equipment_id?: string;
  status?: string;
  priority?: string;
}

export const fetchRecommendations = async (params?: RecommendationFilterParams): Promise<RecommendationResponse[]> => {
  const response = await apiClient.get<RecommendationResponse[]>('/recommendations', { params });
  return response.data;
};

export const generateRecommendations = async (): Promise<RecommendationGenerationSummary> => {
  const response = await apiClient.post<RecommendationGenerationSummary>('/recommendations/generate');
  return response.data;
};

export const updateRecommendationStatus = async (
  recommendationId: string,
  status: 'ACCEPTED' | 'DISMISSED' | 'PENDING'
): Promise<RecommendationResponse> => {
  const response = await apiClient.patch<RecommendationResponse>(
    `/recommendations/${encodeURIComponent(recommendationId)}`,
    { status }
  );
  return response.data;
};
